### DISCUSSION TITLE
response status is 403 /openapi.json

### DISCUSSION THREAD

**Discussion by anhquyetnguyen** [fastapi/fastapi#10431] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi import FastAPI, UploadFile, File, Request, Response
from fastapi.openapi.docs import get_swagger_ui_html
from fastapi.staticfiles import StaticFiles

app = FastAPI()
@app.get("/")
async def root():
    return {"message": "Hello World"}

@app.get("/docs",include_in_schema=False)
async def get_documentation():
    return get_swagger_ui_html(openapi_url="/openapi.json", title="docs")
    
if __name__ == "__main__":
    app.run(debug=False, host="0.0.0.0", port=5000)
```

### Description

When I deploy service on ECS then I access url swagger UI, I have an error

### Operating System

Linux

### FastAPI Version

0.98.0

### Pydantic Version

2.4.2

### Python Version

3.8.0

### Additional Context

![image](https://github.com/tiangolo/fastapi/assets/23186613/bb2e7041-562f-43db-9184-f8b31505a04e)

**Comment by WilliamStam** [comment]
can you open `127.0.0.1:5000/openapi.json` in your browser? something seems to be blocking that. (the swagger docs import the openapi.json file to render stuff)

**Comment by ddroberts77** [comment]
You can secure your FastAPI OpenAPI documentation by disabling the default docs endpoints and implementing custom ones with API key validation. Here's a complete example that shows how to set up protected documentation endpoints using APIKeyHeader and custom route handlers. The key parts are disabling default endpoints (docs_url=None, redoc_url=None, openapi_url=None), implementing API key validation, and creating custom /docs and /openapi.json endpoints that require authentication.


from fastapi import FastAPI, Depends, HTTPException, Security, status
from fastapi.security import APIKeyHeader
from fastapi.openapi.docs import get_swagger_ui_html
from fastapi.openapi.utils import get_openapi
from typing import Optional

# API Key security scheme
api_key_header = APIKeyHeader(name="X-API-Key", auto_error=False)

app = FastAPI(docs_url=None, redoc_url=None, openapi_url=None)

# Your API key validation logic
async def get_api_key(api_key: str = Security(api_key_header)) -> str:
    """Validate API key and return it if valid."""
    if api_key == "your-secret-key":  # Replace with your actual validation
        return api_key
    raise HTTPException(
        status_code=status.HTTP_403_FORBIDDEN,
        detail="Could not validate API Key"
    )

# Custom OpenAPI endpoint with security
@app.get("/openapi.json", include_in_schema=False)
async def get_open_api_endpoint(api_key: str = Depends(get_api_key)):
    return get_openapi(
        title="Your API Title",
        version="1.0.0",
        description="Your API Description",
        routes=app.routes,
    )

# Custom docs endpoint with security
@app.get("/docs", include_in_schema=False)
async def get_documentation(api_key: str = Depends(get_api_key)):
    return get_swagger_ui_html(
        openapi_url="/openapi.json",
        title="API Documentation",
    )

# Example protected endpoint
@app.get("/protected")
async def protected_route(api_key: str = Depends(get_api_key)):
    return {"message": "Access granted to protected route"}

"""
To use this API:
1. Send requests with header: X-API-Key: your-secret-key
2. Access docs at /docs with the same header
3. Access OpenAPI schema at /openapi.json with the same header

Example curl commands:
# Access docs
curl -H "X-API-Key: your-secret-key" http://localhost:8000/docs

# Access OpenAPI schema
curl -H "X-API-Key: your-secret-key" http://localhost:8000/openapi.json

# Access protected endpoint
curl -H "X-API-Key: your-secret-key" http://localhost:8000/protected
"""
